### DISCUSSION TITLE
Description, notes, and metadata...

### DISCUSSION THREAD

**Discussion by pelted** [cooklang/spec#46] [Q&A]
Just found out about this project from a podcast I was listening to the other day and have spent a few hours now moving my markdown and notes-based recipes over here.

I really like the idea in #26, this would help a lot, but I'm still stuck on the best way to handle something like a description and maybe notes for an overall recipe. I find these very helpful in several types of recipes. A description is common, and notes about pairings, or alternatives would also be great. 

As a developer, the comment syntax is a no-brainer, but I really expected these to be visible in the iOS app. It seems like taking a note/comment and having that available to the cook could be valuable. (Umm, I guess what is seen is more a discussion for the app project rather than the spec though.)

**Comment by dubadub** [comment]
@pelted, thanks for your question. As I understood notes aren’t specific to any step and should be displayed on a main recipe screen (before cooking is started), right? It would be really helpful for discussion if you can share an example recipe to demonstrate the use-case.

**Reply by pelted** [reply]
Hey.. yeah I think just having a generic description block of text at the beginning of a recipe would go a long way in support of usability. Sometimes the title just isn't enough. If we could display a paragraph or two of text describing the dish, maybe other dishes or side dishes (even linking to them with brackets) that would go with this current one and why.

As for implementation, maybe something like a divider `-----` or `=====` that could be easily parsed. Everything before this is main recipe details, and paragraphs that follow are steps.

Below is a quick example of a salsa recipe with a divider and a link to another recipe in the first paragraph of the description block. The link would look for a `salsa verde.cook` file and provide a link if found.

```
salsa roja.cook

>> prep time: 5 minutes
>> cook time: 40 minutes
>> difficulty: easy
>> servings: 8
>> tags: mexican, salsa

A versatile red salsa prepared with tomatoes that works well with many Mexican dishes. Also pairs well with the [salsa verde] recipe made with tomatillos.

Make sure that your tomatoes are really red and ripe which gives the salsa better color and a richer flavor. Really fresh cilantro gives the salsa a nice herbal tang. Avoid tired wilted cilantro because it has lost much of its flavor.

Some optional additions, ground roasted pumpkin seeds, or ground cumin may be added to taste

=====

Slice @roma tomatoes{5} in half and @white onion{1} in quarters.

Remove the stems, then slice the @serrano chiles{2} in half and remove the seeds. -- Leave the seeds for more heat.

Add the Roma, onions, serranos, and @garlic{2%cloves} to a #pot with about @water{2%cups}. Bring the water to a boil and then reduce it to low. Simmer ~{20%minutes}.

Place all of the ingredients including the @cilantro{12%stalks} with the cooking water into a #blender. -- Blend in two batches.

Preheat the pot with @cooking oil{2%tbsp} on high. Slowly pour the blended salsa into the hot oil. This is called "seasoning" the salsa and it's an important step to develop the flavor of your salsa.

Reduce the heat and simmer for ~{20%minutes}. Notice that the salsa has reduced and thickened. Add the @salt{1%tsp} and stir. Adjust the salt as needed.
```

**Reply by pelted** [reply]
Another thought around this is that this method could potentially open up the ability to pre-define ingredients in the upper block without changing much else about parsing. It would create flexibility for more quickly adapting recipes to the cooklang format.

For example, a lot of online recipes start with an ingredient list then the steps. [Quick example from www.joshuaweissman.com](https://www.joshuaweissman.com/post/easy-authentic-thai-green-curry)

With an upper description block, you could copy and paste an ingredient list then reformat it to the spec without taking away the ability to also define them in the steps.

```
thai green curry.cook

>> source: https://www.joshuaweissman.com/post/easy-authentic-thai-green-curry

Homemade Thai Green Curry deserves respect when making it. Good thing it's actually very easy and just requires a little elbow grease. 

@coriander seeds{2%tsps}
@cumin seeds{1%tsps}
@white peppercorn{1%tsps}
@serranos{3}
@Thai chilies{9}
@shallots{2}
@lemongrass{1%stalk}
@Galangal{2%inches}
@kaffir lime leaves{3}
@garlic{7%cloves}
@cilantro stems{4%springs}
@lime{1}
@shrimp paste{1/2%tsps}

=====

Pre-heat oven at 350F. Place coriander seeds, cumin seeds, and white peppercorn on a small baking dish; then put into the oven for 5 minutes or lightly toasted. Let them cool down completely. 

In a large #mortar and pestle{}, add spices, grind those until you get a powder, followed by your Thai chilies, and grind those into a paste. Next, add your serrano peppers, grind them down; then add your galangal and keep grinding. (Note.- use your grinder to press it along the wall as you gently twist your pestle.)

Add lemongrass and grind until pestified, about 3 minutes. Next, grind in your kefir lime till smooth, followed by coriander stem until soft, grind shallots until smooth, next grind garlic until smooth, the zest of one lime, and finally shrimp paste, blending all that.
```

**Reply by dubadub** [reply]
Thanks for sharing!

**Reply by dubadub** [reply]
Proposed syntax is brilliant!

**Reply by wkokolis** [reply]
Maybe wrap an introduction/description block with commonly understood block-comment delimiters like `/*` and `*/` or `<--` and `-->` to open/close? 

I love the `=====` between logical sections of the recipe.

**Comment by dubadub** [comment]
Relevant proposal for spec change https://github.com/cooklang/spec/pull/104.

**Comment by dubadub** [comment]
Relevant proposal for spec change https://github.com/cooklang/spec/pull/105.

**Comment by dubadub** [comment]
This is now part of the main specification, closing the discussion.
